### PR TITLE
towards composition-based autoqc - result object changes

### PR DESCRIPTION
**Pull request by mgcam**
- Towards composition-based autoqc:
   - npg_qc::autoqc::results::result - Composition-aware,
     id_run, position and path attributes become optional.
   - npg_qc::autoqc::qc_store - Use DBIx syntax for quering on tag_index
     (future compatibility with planned table schema changes), db load
     compositions of size 1 only.
   - npg_qc::autoqc::db_loader - Result classes support composition,
     but most tables have not been converted yet. Do not generate
     composition if no support in the result table.
   - npg_qc::Schema::ResultSet - Enchancement for a search for tables that
     do not yet support compositions: convert undefined values
     for columns under a unique constraint to database defaults.
 - SeqQC - outdated template for split_stats removed
 - genotype role refactored to remove hardcoded uri encoding

### CODE REVIEW

**Comment by mgcam** [comment]
Run the pipeline for runs 20119 and 20127, output files as expected, json results contain teh composition field. SeqQC view shows these files correctly. The autoqc loader loaded the json autoqc files into teh dev database.

File: lib/npg_qc/autoqc/results/adapter.pm [hunk 1]
```diff
@@ -1,55 +1,30 @@
...
 has reverse_contaminated_read_count => ( is  => 'rw',
                                          isa => 'Maybe[Int]', );
-
-
 has reverse_blat_hash               => ( is  => 'rw',
                                          isa => 'Maybe[HashRef]', );
-
 has forward_start_counts            => ( is  => 'rw',
-                                         isa => 'Maybe[HashRef]', );
+                                        isa => 'Maybe[HashRef]', );
```
Comment (jmtcsngr) [inline lib/npg_qc/autoqc/results/adapter.pm]: Indentation was accidentally lost

File: lib/npg_qc/autoqc/role/result.pm [hunk 2]
```diff
@@ -81,10 +81,61 @@ Human readable check name
...
+
+  my @subsets = $self->composition->component_values4attr('subset');
+  my $subset;
+  if (@subsets) {
+    if (scalar @subsets > 1) {
+      croak 'Multiple subsets within the composition: ' . join q[,] , @subsets;
+    }
+    $subset = $subsets[0];
+  }
+
```
Comment (jmtcsngr) [inline lib/npg_qc/autoqc/role/result.pm]: Consider removing the second validation and add test to the result object

File: lib/npg_qc/autoqc/role/result.pm [hunk 3]
```diff
@@ -102,60 +153,84 @@ Supports comparison on the following attributes:
...
+      croak qq[Value of the $key attribute cannot be compared. Valid attributes: ]
+            . join q[, ], @SEARCH_PARAMETERS;
+	  }
+    if ($key eq q[tag_index]) {
+      if ( !$self->can($key) || (!defined $h->{$key} && defined $self->$key) ||
+          (defined $h->{$key} && !defined $self->$key) ) {
+        return 0;
+      }
+      if (!defined $h->{$key} && !defined $self->$key) { next; }
+	  }
```
Comment (jmtcsngr) [inline lib/npg_qc/autoqc/role/result.pm]: Indented with a <tab>

**Comment by mgcam** [comment]
depends on https://github.com/wtsi-npg/npg_tracking/pull/338

**Comment by mgcam** [comment]
Reverse https://github.com/wtsi-npg/npg_qc/pull/342/commits/554ae098986d18cf5a6bcbfdb4e0dba908936d6a before merging to wtsi/devel

**Pull request merged**